### PR TITLE
Set up project structure and add address type

### PR DESCRIPTION
**Pull request by frrist**
### What
1. Sets up go modules and filecoin-ffi submodule (required by address)
2. Adds a filecoin address type (extracted from lotus)

This PR is just copy-pasta code from lotus.

### CODE REVIEW

File: pkg/types/fi.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		return FIL{}, fmt.Errorf("failed to parse %q as a decimal number", s)
+	}
+
+	r = r.Mul(r, big.NewRat(params.FilecoinPrecision, 1))
+	if !r.IsInt() {
+		return FIL{}, fmt.Errorf("invalid FIL value: %q", s)
+	}
+
+	return FIL{r.Num()}, nil
+}
```
Comment (frrist) [inline pkg/types/fi.go]: used by [bigint_test.go](https://github.com/filecoin-project/go-fil-shared-types/pull/1/files#diff-573fe29b9ce59dcc8760f73d4a5ee957R42)

File: internal/crypto/crypto.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
+package crypto
```
Comment (frrist) [inline internal/crypto/crypto.go:1]: Used in address_test.go. Sits under `internal/` to prevent it from being used outside of this library.

**Review by hannahhoward — APPROVED** [review]
I would say this early on, basic layout is fine, let's merge it...

File: internal/crypto/crypto.go (hunk 2 above)
Comment (anorth) [reply internal/crypto/crypto.go:1]: If this is a copy-paste and any other code than Address might use it, we should make it public and deduplicate between the impls.

Alternatively, change the address test not to use it and remove from this repo.

File: pkg/address/constants.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	if err != nil {
+		panic(err)
+	}
+}
+
+var (
+	// TestAddress is an account with some initial funds in it.
+	TestAddress Address
+	// TestAddress2 is an account with some initial funds in it.
+	TestAddress2 Address
```
Comment (anorth) [inline pkg/address/constants.go:29]: These will introduce coupling between tests, not just in one project but now across multiple repos! Please remove from this repo, the implementations can redefine values if needed.

File: pkg/address/testing.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package address
+
+import (
+	"fmt"
+)
+
+// NewForTestGetter returns a closure that returns an address unique to that invocation.
+// The address is unique wrt the closure returned, not globally.
+func NewForTestGetter() func() Address {
```
Comment (anorth) [inline pkg/address/testing.go:9]: I suggest removing this too. It's not really library code, can easily be re-implemented (possibly differently) by others.

**Review by anorth — COMMENTED** [review]
Thanks! A few notes for a follow-up before we attempt to consume this from elsewhere.

**Pull request merged**